Does not build on centos7 (missing `stdatomic.h` due to gcc <4.9)

https://bugzilla.redhat.com/show_bug.cgi?id=1850174

Failure is:

```
[ghc@1fd8b9c7a733 bytestring-0.11.2.0]$ cabal build
Build profile: -w ghc-8.10.7 -O1
In order, the following will be built (use -v for more details):
 - bytestring-0.11.2.0 (lib) (first run)
Preprocessing library for bytestring-0.11.2.0..
Building library for bytestring-0.11.2.0..

cbits/fpstring.c:42:23: error:
     fatal error: stdatomic.h: No such file or directory
     #include <stdatomic.h>
                           ^
   |
42 | #include <stdatomic.h>
   |                       ^
compilation terminated.
`gcc' failed in phase `C Compiler'. (Exit code: 1)
```

Since the include is guarded anyway, there should be a way to make this build with older gcc's. It seems a configure script would be needed.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -113,3 +113,21 @@ jobs:
           cabal update
           cabal run -w ghc-9.3.20220124 --ghc-options='-fcheck-prim-bounds -fno-ignore-asserts' bytestring-tests
       shell: bash
+
+  old-gcc:
+    runs-on: ubuntu-latest
+    container:
+      image: centos:7
+    steps:
+    - name: install deps
+      run: |
+          yum install -y gcc gmp gmp-devel make ncurses ncurses-compat-libs xz perl
+          curl --proto '=https' --tlsv1.2 -sSf https://get-ghcup.haskell.org | BOOTSTRAP_HASKELL_NONINTERACTIVE=1 sh
+      shell: bash
+    - uses: actions/checkout@v2
+    - name: test
+      run: |
+          source ~/.ghcup/env
+          cabal update
+          cabal run bytestring-tests
+      shell: bash

--- a/cbits/fpstring.c
+++ b/cbits/fpstring.c
@@ -38,8 +38,9 @@
 #include <stdint.h>
 #include <stdbool.h>
 
-#ifndef __STDC_NO_ATOMICS__
+#if defined(__x86_64__) && (__GNUC__ >= 6 || defined(__clang_major__)) && !defined(__STDC_NO_ATOMICS__)
 #include <stdatomic.h>
+#define USE_SIMD_COUNT
 #endif
 
 /* copy a string in reverse */
@@ -114,9 +115,6 @@ size_t fps_count_naive(unsigned char *str, size_t len, unsigned char w) {
     return c;
 }
 
-#if defined(__x86_64__) && (__GNUC__ >= 6 || defined(__clang_major__)) && !defined(__STDC_NO_ATOMICS__)
-#define USE_SIMD_COUNT
-#endif
 
 #ifdef USE_SIMD_COUNT
 __attribute__((target("sse4.2")))

--- a/cbits/is-valid-utf8.c
+++ b/cbits/is-valid-utf8.c
@@ -37,12 +37,14 @@ SUCH DAMAGE.
 #ifdef __x86_64__
 #include <emmintrin.h>
 #include <immintrin.h>
-#include <tmmintrin.h>
 #include <cpuid.h>
-#endif
-
-#ifndef __STDC_NO_ATOMICS__
+#if (__GNUC__ >= 6 || defined(__clang_major__)) && !defined(__STDC_NO_ATOMICS__)
+#include <tmmintrin.h>
 #include <stdatomic.h>
+#else
+// This is needed to support CentOS 7, which has a very old GCC.
+#define CRUFTY_GCC
+#endif
 #endif
 
 #include <MachDeps.h>
@@ -161,7 +163,7 @@ static inline int is_valid_utf8_fallback(uint8_t const *const src, size_t const 
   return 1;
 }
 
-#ifdef __x86_64__
+#if defined(__x86_64__) && !defined(CRUFTY_GCC)
 
 // SSE2
 
@@ -658,7 +660,7 @@ is_valid_utf8_avx2(uint8_t const *const src, size_t const len) {
 
 #endif
 
-#ifdef __x86_64__
+#if defined(__x86_64__) && !defined(CRUFTY_GCC)
 static inline bool has_sse2() {
   uint32_t eax = 0, ebx = 0, ecx = 0, edx = 0;
   __get_cpuid_count(1, 0, &eax, &ebx, &ecx, &edx);
@@ -687,7 +689,7 @@ int bytestring_is_valid_utf8(uint8_t const *const src, size_t const len) {
   if (len == 0) {
     return 1;
   }
-#ifdef __x86_64__
+#if defined(__x86_64__) && !defined(CRUFTY_GCC)
   static _Atomic is_valid_utf8_t s_impl = (is_valid_utf8_t)NULL;
   is_valid_utf8_t impl = atomic_load_explicit(&s_impl, memory_order_relaxed);
   if (!impl) {

